### PR TITLE
Fixes #7222 - Added before_validation handle that will merge restrictions from the current user

### PR DESCRIPTION
**Pull request by ShimShtein**

### CODE REVIEW

File: app/models/filter.rb [hunk 1]
```diff
@@ -180,4 +180,46 @@ def allowed_taxonomies
       errors.add(:location_ids, _('You can\'t assign locations to this resource'))
     end
   end
+
+  def fix_to_current_user_permissions
+    return unless role_id
+    return if permission_ids.empty?
+    user = User.current
+    return unless user
+    return if user.admin?
```
Comment (shlomizadok) [inline app/models/filter.rb]: You can do the four lines above in two

``` ruby
user = User.current
return unless role_id || user
return if permission_ids.empty? || user.admin?
```

File: app/models/filter.rb [hunk 2]
```diff
@@ -180,4 +180,46 @@ def allowed_taxonomies
...
+      errors.add(:resource_type, (_('Current user is not allowed to resource type: %s') % new_resource_type))
+      return
+    end
+
+    # allow adding filters to permitted resources only
+    allowed_permissions =  new_permissions & current_permissions[new_resource_type].permissions
+    self.permission_ids = allowed_permissions.map(&:id)
+
+    # restrict filter condition if the user has custom filter
+    new_search = search || ''
```
Comment (shlomizadok) [inline app/models/filter.rb]: `search ||= '' ?`

File: app/models/filter.rb (hunk 1 above)
Comment (tbrisker) [inline app/models/filter.rb]: how about something like:

``` ruby
def need_to_fix_permissions
  role_id.present? && permission_ids.present? && User.current.present? && !User.current.admin?
end
```

and then only [conditionally](http://guides.rubyonrails.org/active_record_callbacks.html#using-if-and-unless-with-a-symbol) run this callback if the condition is true?

File: app/models/filter.rb [hunk 3]
```diff
@@ -180,4 +180,46 @@ def allowed_taxonomies
...
+      unless new_search == current_permissions[new_resource_type].search
+        #new_search can be 0: '' 1: 'new_condition' 2:'current_condition and (new_condition)'
+        current_condition = current_permissions[new_resource_type].search
+        if new_search.empty?
+          # just copy my search [0]
+          self.search = current_condition
+          # check if the new condition does not begin with old one (not augmented)
+        elsif new_search_tokens.take(current_search_tokens.size) != current_search_tokens
+          current_condition << " and ("
+          self.search = "#{current_condition}#{new_search})"
```
Comment (ares) [inline app/models/filter.rb]: I don't think we should change the data like this. I understand that you're trying to make it easier for the user but I think we should not save the filter and display an error in this case ideally suggesting what conditions must be present. The reason is that user should know what's the resulting filter.

**Comment by ares** [comment]
Before you invest too much time into this issue, I'd recommend asking why it's necessary. I think that if you granted user A permissions to grant permissions to other users, you wouldn't limit user A on other resource. Maybe only by organization/location which works today. I'm not sure if it makes sense to have user A being able to see only architecture x86_64 and let her edit roles at the same time. I know it's valid setup now, so maybe we should prevent it or display a warning that user that can modify roles and filters can assign more permissions to themselves and it's pointless to try to limit their resources. Otherwise this would imply that view_architecture has impact on edit_filter. So far we don't have any relation between filters and if we need to introduce it, I think it should be designed first.

**Comment by ShimShtein** [comment]
I have updated the PR.
After a short discussion with @thomasmckay, it seems like the feature is not very pressing. The main concern was for creating an organization admin.
Without the PR any user that has permissions to roles/filters is effectively an admin (scoped for org/loc) because he can add any permission to his own user.
If the PR gets downvotes, I'll close it.

**Comment by dLobatog** [comment]
I'd go with @ares 's answer here and close as nofix - the value isn't worth the added complexity IMO. A warning might be appropriate though, but I think that if you are assigning role/filter create permissions to an user, and they can modify their role, you should be aware they're basically able to escalate to admin.

**Comment by ohadlevy** [comment]
@thomasmckay your thoughts ^^ ?

**Comment by thomasmckay** [comment]
Yes, closing this PR and the associated issue is good. The topic of creating a true org admin, versus the current checkbox-admin, is one that will require a design phase. It is necessary for true multitenant behavior which has not been accepted upstream as a requirement at this time.

**Comment by ShimShtein** [comment]
Closing this PR, since the issue has been rejected.

**Pull request closed without merging**